src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/SuppressMessageAttributeState.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+#if NETSTANDARD2_0
+using Roslyn.Utilities;
+#endif
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal partial class SuppressMessageAttributeState
+    {
+        internal const string SuppressMessageScope = "Scope";
+        internal const string SuppressMessageTarget = "Target";
+
+        private static readonly ImmutableDictionary<string, TargetScope> s_targetScopesMap = CreateTargetScopesMap();
+        private static readonly ObjectPool<List<ISymbol>> s_listPool = new ObjectPool<List<ISymbol>>(() => new List<ISymbol>());
+
+        private readonly Compilation _compilation;
+        private readonly INamedTypeSymbol _suppressMessageAttributeType;
+        private readonly Lazy<ImmutableDictionary<SyntaxNode, AttributeData>> _lazySuppressMessageAttributesBySyntax;
```

Comment:
❔ Why do we need this field? It seems very expensive to compute, even though we only need to look at the attributes individually/incrementally.

Reply:
Without this field, for every attribute syntax callback, we need to walk through all the global attributes on compilation's assembly symbol and all module symbols to match which `AttributeData` has a matching syntax for a given attribute syntax. This would be computationally very expensive. The root issue here is there is no API that allows mapping a given attribute syntax to `AttributeData` which has the semantic information associated with the attribute binding. Compiler should either provide an API `public AttributeData SemanticModel.GetAttributeInfo(SyntaxNode attributeSyntax)` and/or would add IOperation support for attributes (https://github.com/dotnet/roslyn/issues/18198) and this would no longer be required. I will file a separate issue and tag relevant compiler devs.

Reply:
Why not just check the parent syntax nodes?

Reply:
Filed https://github.com/dotnet/roslyn/issues/44340

Reply:
📝 I've seen assembly attribute walks in many performance traces. I expect this requirement to be eliminated before merge, or if there is no way to do it we should review the implications on typing scenarios in details before moving forward.

Reply:
> Why not just check the parent syntax nodes?

Not sure I understand. How would we get the required `AttributeData` for semantic analysis?

Reply:
Had an offline discussion with @sharwell and we agree on concerns from typing in global suppressions files with this approach. I'll experiment with Operation action with special detection (`OperationKind.None` nodes with an `AttributeSyntax`) as a workaround for #18198 to workaround this issue.

Reply:
Now removed with IOperation based workaround.